File: project_euler/02/029.go
// Copyright [2017] <dmnsn7@gmail.com>

package main

import "fmt"

const cA = 100
const cB = 100

type pair struct {
	fisrt  int
	second int
}

func main() {
	isVisited := map[pair]int{}
	cnt := (cA - 1) * (cB - 1)
	for i := 2; i <= cA; i++ {
		for j := 2; j <= cB; j++ {
			ti, minDiv := i, 1
			for k := 2; k <= ti/k; k++ {
				if ti%k == 0 {
					minDiv *= k
					for ti%k == 0 {
						ti /= k
					}
				}
			}

			tti, level := i, 0
			for minDiv != 1 && tti%minDiv == 0 {
				tti /= minDiv
				level++
			}

			if tti != 1 {
				isVisited[pair{i, j}] = 0
			} else if _, exist := isVisited[pair{minDiv, level * j}]; !exist {
				isVisited[pair{minDiv, level * j}] = 0
			} else {
				cnt--
			}
		}
	}

	fmt.Printf("%d\n", cnt)
}

File: project_euler/02/029.py
#!/usr/bin/env python3
'''Copyright [2017] <dmnsn7@gmail.com>'''

A = 100
B = 100


def main():
    is_visited = set()
    cnt = (A - 1) * (B - 1)
    for i in range(2, A + 1):
        for j in range(2, B + 1):
            tmp_i, min_div = i, 1
            for k in range(2, tmp_i + 1):
                if k > tmp_i / k:
                    break
                if tmp_i % k == 0:
                    min_div *= k
                    while tmp_i % k == 0:
                        tmp_i /= k

            tmp_tmp_i, level = i, 0
            while min_div != 1 and tmp_tmp_i % min_div == 0:
                tmp_tmp_i /= min_div
                level += 1

            if tmp_tmp_i != 1:
                is_visited.add((i, j))
            elif (min_div, level * j) not in is_visited:
                is_visited.add((min_div, level * j))
            else:
                cnt -= 1

    print(cnt)


if __name__ == '__main__':
    main()
